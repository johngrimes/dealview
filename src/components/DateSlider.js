import React from 'react'
import PropTypes from 'prop-types'
import ReactSlider from 'react-slider'
import throttle from 'lodash.throttle'

import './styles/DateSlider.css'

class DateSlider extends React.Component {
  static propTypes = {
    dates: PropTypes.arrayOf(PropTypes.string).isRequired,
    selected: PropTypes.string,
    initialised: PropTypes.bool.isRequired,
    className: PropTypes.string,
    onChange: PropTypes.func,
  }

  constructor(props) {
    super(props)

    const selectedIndex = props.dates.indexOf(props.selected)
    this.state = {
      selected: props.selected,
      selectedIndex: selectedIndex !== -1 ? selectedIndex : 0,
    }

    this.handleChange = this.handleChange.bind(this)
  }

  handleChangeUnthrottled(value) {
    const selectedIndex = parseInt(value, 10)
    const selected = this.props.dates[selectedIndex]
    this.setState(
      () => ({ selectedIndex, selected }),
      () => {
        if (this.props.onChange) this.props.onChange(selected)
      }
    )
  }
  handleChange = throttle(this.handleChangeUnthrottled, 100)

  componentWillReceiveProps(props) {
    const selectedIndex = props.dates.indexOf(props.selected)
    this.setState(() => ({
      selected: props.selected,
      selectedIndex: selectedIndex !== -1 ? selectedIndex : 0,
    }))
  }

  render() {
    const { dates, className, initialised } = this.props
    const { selected, selectedIndex } = this.state
    return (
      <div
        className={
          initialised ? 'date-slider' : 'date-slider date-slider-uninitialised'
        }
      >
        <div className='date-tag start-date'>
          {dates[0]}
        </div>
        <div className='date-tag end-date'>
          {dates[dates.length - 1]}
        </div>
        <ReactSlider
          className={className ? `slider ${className}` : className}
          min={0}
          max={dates.length - 1}
          defaultValue={0}
          value={selectedIndex}
          onChange={this.handleChange}
        >
          <div className='date-tag selected-date'>
            {selected}
          </div>
        </ReactSlider>
      </div>
    )
  }
}

export default DateSlider
