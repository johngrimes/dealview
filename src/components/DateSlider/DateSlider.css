.slider {
  width: 500px;
  margin: 20px 0 20px;
  display: block;
}
