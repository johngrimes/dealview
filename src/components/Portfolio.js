// @flow

import React from 'react'
import DatePicker from 'react-datepicker'
import moment from 'moment'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import type { Dispatch } from 'redux'

import { loadBalanceSheet, updateBalanceSheet } from 'actions/balanceSheet'
import { DateFormat, formatDollars } from 'types/commonTypes'
import type { GlobalState } from 'store'
import type { AssetState } from 'reducers/assets'
import type { BalanceSheetState } from 'reducers/balanceSheet'

type Props = {
  assets: AssetState,
  balanceSheet: BalanceSheetState,
  dispatch: Dispatch
}
type State = {
  date: string,
}

class Portfolio extends React.Component {
  props: Props
  state: State

  constructor(props: Props) {
    super(props)
    if (this.props.balanceSheet.fresh === false) {
      this.props.dispatch(updateBalanceSheet(this.props.assets.objects, {}, '2017-05-28', '2027-05-28'))
    } else if (this.props.balanceSheet.status === 'uninitialised') {
      this.props.dispatch(loadBalanceSheet())
    }
    this.state = { date: '2018-05-28' }
  }

  render() {
    const date = moment(this.state.date)
    const balanceSheet = typeof this.props.balanceSheet.balanceSheet === 'object'
      ? this.props.balanceSheet.balanceSheet[this.state.date]
      : undefined

    console.log('this.props.balanceSheet.balanceSheet', this.props.balanceSheet.balanceSheet)
    console.log('balanceSheet', balanceSheet)
    return (
      <div className='portfolio'>
        <DatePicker showYearDropdown dateFormat={DateFormat} selected={date} onChange={moment => this.setState({ date: moment.format(DateFormat) })} />
        <div className='assets'>
          <Link to='/portfolio/assets'>Assets</Link>
          <span className='assets-total'>{balanceSheet ? formatDollars(balanceSheet.totalAssets) : '?'}</span>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state: GlobalState) => {
  return {
    assets: state.assets,
    balanceSheet: state.balanceSheet,
  }
}

export default connect(mapStateToProps)(Portfolio)
