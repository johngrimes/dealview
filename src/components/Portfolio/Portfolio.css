.portfolio .slider-container {
  padding: 0 47px;
}

.portfolio .date-slider {
  width: 100%;
}
