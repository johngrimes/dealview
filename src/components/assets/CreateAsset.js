// @flow

import React from 'react'
import type { Children } from 'react'

const CreateAsset = (props: { children: Children }) => {
  const { children } = props

  return (
    <div className='create-asset'>
      <h1>Create Asset</h1>
      {children}
    </div>
  )
}

export default CreateAsset
