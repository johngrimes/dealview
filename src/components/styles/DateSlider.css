.slider {
  clear: both;
  margin: 34px 0 40px;
  height: 8px;
  background: linear-gradient(#aaa, #ddd);
  display: block;
  border-radius: 6px;
}

.handle {
  background: linear-gradient(#129fea, #216ba5);
  width: 20px;
  height: 20px;
  border-radius: 10px;
  box-sizing: border-box;
  box-shadow: 0px 1px 5px 1px #aaa;
  position: relative;
  top: -6px;
}

.date-tag {
  background: transparent url('../images/date-tag.svg') no-repeat top left;
  width: 76px;
  background-size: 76px;
  font-size: 13px;
  text-align: center;
  height: 30px;
}

.start-date, .end-date {
  margin-top: 10px;
  margin-bottom: -6px;
  position: relative;
}

.start-date {
  float: left;
  left: -36px;
}

.end-date {
  float: right;
  left: 36px;
}

.selected-date {
  background: transparent url('../images/date-tag-sel.svg') no-repeat top left;
  line-height: 28px;
  position: relative;
  left: -28px;
  top: 21px;
}

.date-slider-uninitialised .date-tag,
.date-slider-uninitialised .handle {
  display: none;
}
