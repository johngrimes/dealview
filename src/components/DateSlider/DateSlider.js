// @flow

import React from 'react'
import ReactSlider from 'react-slider'

import 'components/DateSlider/DateSlider.css'

type Props = {
  +dates: string[],
  +className?: string,
  +onChange?: (value: string) => void,
}
type State = {
  +selected: string,
  +selectedIndex: number,
}

class DateSlider extends React.Component {
  props: Props
  state: State
  handleChange: (value: string) => void

  constructor(props: Props) {
    super(props)
    this.state = {
      selectedIndex: 0,
      selected: props.dates[0],
    }

    this.handleChange = this.handleChange.bind(this)
  }

  handleChange(value: string): void {
    const selectedIndex = parseInt(value)
    const selected = this.props.dates[selectedIndex]
    this.setState(
      () => ({ selectedIndex, selected }),
      () => { if (this.props.onChange) this.props.onChange(selected) }
    )
  }

  render() {
    const { dates, className } = this.props
    const { selected } = this.state
    return (
      <div>
        <div className='date-tag start-date'>{dates[0]}</div>
        <div className='date-tag end-date'>{dates[dates.length - 1]}</div>
        <ReactSlider className={className ? `slider ${className}` : className}
          min={0} max={dates.length - 1} defaultValue={0}
          onChange={this.handleChange}>
          <div className='date-tag selected-date'>{selected}</div>
        </ReactSlider>
      </div>
    )
  }
}

export default DateSlider
