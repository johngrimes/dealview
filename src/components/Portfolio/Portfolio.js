// @flow

import React from 'react'
import DatePicker from 'react-datepicker'
import moment from 'moment'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'

import Breadcrumbs from 'components/Breadcrumbs/Breadcrumbs'
import { loadBalanceSheet, updateBalanceSheet } from 'actions/balanceSheet'
import { DateDisplayFormat, DateStorageFormat, formatDollars } from 'types/commonTypes'
import DateSlider from 'components/DateSlider/DateSlider'
import type { GlobalState } from 'store'
import type { AssetState } from 'reducers/assets'
import type { BalanceSheetState } from 'reducers/balanceSheet'
import type { BreadcrumbTrail } from 'components/Breadcrumbs/Breadcrumbs'

import 'components/Portfolio/Portfolio.css'

type Props = {
  +assets: AssetState,
  +balanceSheet: BalanceSheetState,
  +dispatch: any
}
type State = {
  +date: string,
}

class Portfolio extends React.Component {
  props: Props
  state: State
  handleDateChange: (date: string) => void

  constructor(props: Props) {
    super(props)
    if (this.props.balanceSheet.fresh === false) {
      this.props.dispatch(updateBalanceSheet(this.props.assets.objects, {}, '2017-05-28', '2027-05-28'))
    } else if (this.props.balanceSheet.status === 'uninitialised') {
      this.props.dispatch(loadBalanceSheet())
    }
    this.state = { date: '2018-05-28' }

    this.handleDateChange = this.handleDateChange.bind(this)
  }

  breadcrumbs(): BreadcrumbTrail {
    return [
      { display: 'Portfolio', path: '/portfolio' },
    ]
  }

  handleDateChange(date: string): void {
    this.setState(() => ({ date }))
  }

  render() {
    const date = moment(this.state.date)
    const balanceSheet = typeof this.props.balanceSheet.balanceSheet === 'object'
      ? this.props.balanceSheet.balanceSheet[this.state.date]
      : undefined
    const dates = typeof this.props.balanceSheet.balanceSheet === 'object'
      ? Object.keys(this.props.balanceSheet.balanceSheet)
      : []

    return (
      <div className='portfolio'>
        <Breadcrumbs breadcrumbs={this.breadcrumbs()} />
        <div className='slider-container'>

          <DateSlider dates={dates} className='date-slider' onChange={this.handleDateChange} />
        </div>
        <DatePicker showYearDropdown dateFormat={DateDisplayFormat} selected={date} onChange={moment => this.setState(() => ({ date: moment.format(DateStorageFormat) }))} />
        <div className='assets'>
          <Link to='/portfolio/assets'>Assets</Link>
          <span className='assets-total'>{balanceSheet ? formatDollars(balanceSheet.totalAssets) : '?'}</span>
        </div>
        <div className='liabilities'>
          <Link to='/portfolio/liabilities'>Liabilities</Link>
          <span className='assets-total'>{balanceSheet ? formatDollars(balanceSheet.totalLiabilities) : '?'}</span>
        </div>
        <div className='equity'>
          <span>Equity</span>
          <span className='equity-total'>{balanceSheet ? formatDollars(balanceSheet.equity) : '?'}</span>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state: GlobalState) => {
  return {
    assets: state.assets,
    balanceSheet: state.balanceSheet,
  }
}

export default connect(mapStateToProps)(Portfolio)
