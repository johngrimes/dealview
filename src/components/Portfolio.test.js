import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'
import MockDate from 'mockdate'

import { Portfolio } from './Portfolio.js'
import DateSlider from './DateSlider.js'
import { validBalanceSheet1 } from '../data/fixtures/balanceSheet.js'
import { validAssets } from '../data/fixtures/asset.js'
import { validLiabilities } from '../data/fixtures/liability.js'
import { DateStorageFormat } from '../data/commonTypes.js'
import { InitialBalanceSheetState } from '../reducers/balanceSheet.js'
import { InitialAssetState } from '../reducers/assets.js'

describe('Portfolio', () => {
  const props = {
    balanceSheet: {
      status: 'loaded',
      fresh: true,
      balanceSheet: validBalanceSheet1,
    },
    assets: {
      status: 'loaded',
      objects: validAssets,
    },
    liabilities: {
      status: 'loaded',
      objects: validLiabilities,
    },
  }

  beforeEach(() => {
    MockDate.set('2017-01-01')
  })

  it('should render', () => {
    const wrapper = shallow(<Portfolio {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should pass correct props to DateSlider', () => {
    const wrapper = shallow(<Portfolio {...props} />)
    const dateSlider = wrapper.find(DateSlider)
    expect(dateSlider.prop('dates')).toEqual(Object.keys(validBalanceSheet1))
    expect(dateSlider.prop('selected')).toBe(moment().format(DateStorageFormat))
  })

  it('should pass correct props to DateSlider when balance sheet updates', () => {
    const initialProps = {
      balanceSheet: InitialBalanceSheetState,
      assets: InitialAssetState,
      dispatch: jest.fn(),
    }
    const wrapper = shallow(<Portfolio {...initialProps} />)
    wrapper.setProps(props)
    const dateSlider = wrapper.find(DateSlider)
    expect(dateSlider.prop('dates')).toEqual(Object.keys(validBalanceSheet1))
    expect(dateSlider.prop('selected')).toBe(moment().format(DateStorageFormat))
  })

  it('should update state when date changes', () => {
    const wrapper = shallow(<Portfolio {...props} />)
    const dateSlider = wrapper.find(DateSlider)
    dateSlider.prop('onChange')('2017-06-06')
    expect(wrapper.state('date')).toEqual('2017-06-06')
  })

  afterEach(() => {
    MockDate.reset()
  })
})
