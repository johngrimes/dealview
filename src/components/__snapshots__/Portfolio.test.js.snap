// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio should render 1`] = `
<div
  className="portfolio"
>
  <Breadcrumbs
    breadcrumbs={
      Array [
        Object {
          "display": "Portfolio",
          "path": "/portfolio",
        },
      ]
    }
  />
  <BalanceSheetChart
    balanceSheet={
      Object {
        "2017-01-01": Object {
          "equity": 155070,
          "totalAssets": 1000070,
          "totalLiabilities": 845000,
        },
        "2017-01-02": Object {
          "equity": 165080,
          "totalAssets": 1000080,
          "totalLiabilities": 835000,
        },
      }
    }
  />
  <div
    className="slider-container"
  >
    <DateSlider
      className="date-slider"
      dates={
        Array [
          "2017-01-01",
          "2017-01-02",
        ]
      }
      initialised={true}
      onChange={[Function]}
      selected="2017-01-01"
    />
  </div>
  <div
    className="balance-sheet"
  >
    <div
      className="assets"
    >
      <Link
        replace={false}
        to="/portfolio/assets"
      >
        Assets
      </Link>
      <span
        className="assets-total"
      >
        $1,000,070
      </span>
    </div>
    <div
      className="liabilities"
    >
      <Link
        replace={false}
        to="/portfolio/liabilities"
      >
        Liabilities
      </Link>
      <span
        className="liabilities-total"
      >
        $845,000
      </span>
    </div>
    <div
      className="equity"
    >
      <span>
        Equity
      </span>
      <span
        className="equity-total"
      >
        $155,070
      </span>
    </div>
  </div>
</div>
`;
